src/rules/match.c

```diff
@@ -0,0 +1,465 @@
+    if (!item->kobj) {
+      return 0;
+    }
+  }
+  int ret = 0;
+  RedisModule_HashGet(item->kobj, REDISMODULE_HASH_EXISTS, hrule->field, &ret, NULL);
+  return ret;
+}
+
+void SchemaCustomCtx_Index(SchemaCustomCtx *ctx, IndexSpec *spec, IndexItemAttrs *attrs) {
+  MatchAction *action = actionForIndex(spec, ctx->action);
+  action->attrs = *attrs;
+}
+
+MATCHFUNC(matchCustom) {
+  SchemaCustomRule *crule = (SchemaCustomRule *)r;
+  SchemaCustomCtx cc = {.action = results};
+  return crule->check(ctx, item, crule->arg, &cc);
+}
+
+void SchemaRule_Free(SchemaRule *r) {
+  switch (r->rtype) {
+    case SCRULE_TYPE_EXPRESSION: {
+      SchemaExprRule *serule = (SchemaExprRule *)r;
+      RSExpr_Free(serule->exprobj);
+      RLookup_Cleanup(&serule->lk);
+      rm_free(serule->exprstr);
+      break;
+    }
+    case SCRULE_TYPE_HASFIELD: {
+      SchemaHasFieldRule *hfrule = (SchemaHasFieldRule *)r;
+      rm_free(hfrule->field);
+      break;
+    }
+    case SCRULE_TYPE_KEYPREFIX: {
+      SchemaPrefixRule *prule = (SchemaPrefixRule *)r;
+      rm_free(prule->prefix);
+      break;
+    }
+    case SCRULE_TYPE_MATCHALL:
+    case SCRULE_TYPE_CUSTOM:
+      break;
+  }
+
+  SchemaAction *action = &r->action;
+  switch (action->atype) {
+    case SCACTION_TYPE_GOTO:
+      rm_free(action->u.goto_);
+      break;
+    case SCACTION_TYPE_LOADATTR:
+      SCAttrFields_Decref(action->u.lattr);
+      action->u.lattr = NULL;
+      break;
+    case SCACTION_TYPE_SETATTR:
+    case SCACTION_TYPE_ABORT:
+    case SCACTION_TYPE_INDEX:
+    case SCACTION_TYPE_CUSTOM:
+      break;
+  }
+
+  if (r->rawrule) {
+    size_t n = array_len(r->rawrule);
+    for (size_t ii = 0; ii < n; ++ii) {
+      rm_free(r->rawrule[ii]);
+    }
+    array_free(r->rawrule);
+  }
+  if (r->spec) {
+    IndexSpec_Decref(r->spec);
+  }
+  rm_free(r->name);
+  rm_free(r);
+}
+
+int SchemaRules_AddArgsInternal(SchemaRules *rules, IndexSpec *spec, const char *name,
+                                ArgsCursor *ac, QueryError *err) {
+  // First argument is the name...
+  size_t beginpos = AC_Tell(ac);
+  const char *rtype = NULL;
+  int rc = AC_GetString(ac, &rtype, NULL, 0);
+  if (rc != AC_OK) {
+    QueryError_SetError(err, QUERY_EPARSEARGS, "Missing type for rule");
+    return REDISMODULE_ERR;
+  }
+
+  SchemaRule *r = NULL;
+  if (!strcasecmp(rtype, "PREFIX")) {
+    r = parsePrefixRule(ac, err);
+  } else if (!strcasecmp(rtype, "EXPR")) {
+    r = parseExprRule(ac, err);
+  } else if (!strcasecmp(rtype, "HASFIELD")) {
+    r = parseHasfieldRule(ac, err);
+  } else if (!strcasecmp(rtype, "*")) {
+    r = parseWildcardRule(ac, err);
+  } else {
+    QueryError_SetErrorFmt(err, QUERY_ENOOPTION, "No such match type `%s`\n", rtype);
+    return REDISMODULE_ERR;
+  }
+
+  if (!r) {
+    return REDISMODULE_ERR;
+  }
+  const char *astr = NULL;
+  if (AC_GetString(ac, &astr, NULL, 0) != AC_OK) {
+    astr = "INDEX";
```

Comment:
Why not return error?
In `extractAction` "INDEX" is checked and is a condition to parse SETATTR and LOAD ATTR